### PR TITLE
Cache in filecache

### PR DESCRIPTION
**Pull request by butonic**
alternative approach to https://github.com/owncloud/core/pull/28165

### CODE REVIEW

File: lib/private/Files/Cache/Cache.php [hunk 1]
```diff
@@ -79,6 +80,11 @@ class Cache implements ICache {
 	protected $connection;
 
 	/**
+	 * @var CappedMemoryCache
+	 */
+	public static $metaDataCache = null;
```
Comment (DeepDiver1975) [inline lib/private/Files/Cache/Cache.php]: why static?

File: lib/private/Files/Cache/Cache.php [hunk 2]
```diff
@@ -137,9 +155,10 @@ public function get($file) {
 		//merge partial data
 		if (!$data and is_string($file)) {
 			if (isset($this->partial[$file])) {
-				$data = $this->partial[$file];
+				return $this->partial[$file]; // return here, no need cache partial entries
```
Comment (DeepDiver1975) [inline lib/private/Files/Cache/Cache.php]: are we sure this is not necessary?

File: lib/private/Files/Cache/Wrapper/CacheIndex.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * Jail to a subdirectory of the wrapped cache
+ */
+class CacheIndex extends CacheWrapper {
+
+	/**
+	 * @var CappedMemoryCache
+	 */
+	private static $metaDataCache = null;
```
Comment (DeepDiver1975) [inline lib/private/Files/Cache/Wrapper/CacheIndex.php]: static ?

**Review by DeepDiver1975 — DISMISSED** [review]

**Comment by PVince81** [comment]
@mrow4a would be good if you could also check this out. I remember you looked into wasteful DB queries related to file cache. Maybe you have more insights. Didn't you have a PR adressing this or some piece of this differently ?

File: lib/private/Files/Cache/Wrapper/CacheIndex.php (hunk 3 above)
Comment (butonic) [inline lib/private/Files/Cache/Wrapper/CacheIndex.php]: um leftover code from the previous attempt.

File: lib/private/Files/Cache/Cache.php (hunk 1 above)
Comment (butonic) [inline lib/private/Files/Cache/Cache.php]: testing revealed that we instantiate more than one instance of this class which works against the caching effect.

**Comment by mrow4a** [comment]
Will have a look also later, looks really promising, it is something I wanted to have. There is one disavantage -> if someone forgets to use CacheIndex somewhere and uses filecache directly -> :boom:

**Comment by mrow4a** [comment]
Any new comments? I would really see sth like that be merged, is small filecache code refactor and protection needed?

**Comment by PVince81** [comment]
I'm not sure. It feels like a workaround on top of a sub-par architecture.

On the other hand, caching stuff locally is what we've been doing everywhere with other things to prevent talking to the DB repeatedly.

**Comment by felixboehm** [comment]
moved to triage, assuming this will not be ready for 10.0.3

**Comment by tomneedham** [comment]
Rebased

**Comment by butonic** [comment]
encryption needs also needs changes: https://github.com/owncloud/encryption/pull/13

**Comment by tomneedham** [comment]
8.2 -> 10.0.3beta2 upgrade succeeded with this - which will have helped speed up the upgrade process for large amounts of reshares when coming from 8.2.

**Comment by PVince81** [comment]
`metaDataCache` should rather be an instance variable of our default Cache implementation.

Custom cache implementations would not benefit from this and might not even extend from our `Cache` implementations.

The only advantage of having it static here is the ability to quickly clear it from various locations, while normally you should go to the storage and do a `$storage->getCache()` which is usually more expensive.

**Comment by PVince81** [comment]
Also I wonder if this should/could be implemented as a CacheWrapper instead. We might lose the ability to clear it from outside unless we keep the staticness.

File: lib/private/Files/Cache/Cache.php [hunk 4]
```diff
@@ -187,24 +196,44 @@ public function getFolderContentsById($fileId) {
...
+				self::$metaDataCache->set($getPathByIdkey, $data['path']);
+				return $entry;
 			}, $files);
 		} else {
 			return [];
 		}
 	}
 
+	private function fixTypes(array &$data) {
+		$data['fileid'] = 0 + $data['fileid'];
```
Comment (PVince81) [inline lib/private/Files/Cache/Cache.php]: do we really need this here for caching ? I expect to see potential additional issues

Comment (PVince81) [reply lib/private/Files/Cache/Cache.php]: never mind, I see this was moved from another place

**Comment by tomneedham** [comment]
Happy to postpone this.

**Comment by butonic** [comment]
@PVince81 The initial implementation was a cache wrapper: https://github.com/owncloud/core/pull/28165/files

Nice and clean approach ... isn't it? Unfortunately, we
1. Instantiate multiple instances for the same share storages, defeating the cache effect.
2. Tests, encryption and other technical debt directly manipulate the filecache in the db by executing their own SQL.

I added the necessary things we need to do: https://github.com/owncloud/core/pull/28166/files#diff-b4012cb6106f634acc5ff12ef395021fR85-R87

so ... no this cannet be an instance variable ... because ... still too much crappy code in other places.

**Comment by PVince81** [comment]
> so ... no this cannot be an instance variable ... because ... still too much crappy code in other places.

Argh, you're right

**Comment by PVince81** [comment]
moving to "planned", let's have this merged first thing for 10.0.4alpha/beta

**Comment by PVince81** [comment]
@butonic revive or defer ?

**Comment by mrow4a** [comment]
I can see one disadvantage - if new code is written which writes to filecache, dev needs to be aware that cache needs to be cleaned with that static function. Apart from that - if no fragment of the code which writes to filecache has not been forgoten - looks good.

File: lib/private/Files/Cache/Cache.php [hunk 5]
```diff
@@ -78,6 +79,18 @@ class Cache implements ICache {
 	 */
 	protected $connection;
 
+	/**
+	 * caches 2048 filecache rows
+	 *
+	 * FIXME first, refactor this class as an entity and add any necessary methods to its mapper
+	 * FIXME then, get rid of all the places that directly touch the tables
+	 * FIXME finally, make this private non-static
```
Comment (mrow4a) [inline lib/private/Files/Cache/Cache.php]: Do you plan it to have as separate PR? Do you expect it to be large change later?

**Review by mrow4a — COMMENTED** [review]
@butonic please see the comments

**Comment by PVince81** [comment]
- [ ] please rebase for more recent CI
- [ ] please run smashbox tests against this to confirm

**Comment by butonic** [comment]
- [ ] move more sql to the filecache class
- [ ] always clear filecache in memory cache after every repair step
- [ ] always clear filecache in memory cache after every background job
- [ ] always clear filecache in memory cache after every test class

**Comment by mrow4a** [comment]
Safer, alternative approach to this PR by fixing the app logic - https://github.com/owncloud/core/pull/31482

**Comment by DeepDiver1975** [comment]
@butonic would you agree on pushing #31486 instead of this one? THX

File: apps/files/tests/BackgroundJob/DeleteOrphanedItemsJobTest.php [hunk 6]
```diff
@@ -220,6 +229,9 @@ public function testClearCommentReadMarks() {
 				'path' => $query->createNamedParameter('apps/files/tests/deleteorphanedtagsjobtest.php'),
 				'path_hash' => $query->createNamedParameter(\md5('apps/files/tests/deleteorphanedtagsjobtest.php')),
 			])->execute();
+		if (Cache::$metaDataCache !== null) {
```
Comment (mrow4a) [inline apps/files/tests/BackgroundJob/DeleteOrphanedItemsJobTest.php]: Clear metadata cache before executing the query.

File: apps/files_sharing/lib/Command/CleanupRemoteStorages.php [hunk 7]
```diff
@@ -129,6 +130,9 @@ public function deleteFiles ($numericId, OutputInterface $output) {
 			);
 		$output->write("deleting files for storage $numericId ... ");
 		$count = $queryBuilder->execute();
+		if (Cache::$metaDataCache !== null) {
```
Comment (mrow4a) [inline apps/files_sharing/lib/Command/CleanupRemoteStorages.php]: clear before executing the query

File: apps/files_sharing/tests/Command/CleanupRemoteStoragesTest.php [hunk 8]
```diff
@@ -101,6 +102,9 @@ protected function setup() {
 					$filesQuery->setParameter(2, \md5('file' . $i));
 					$filesQuery->execute();
 				}
+				if (Cache::$metaDataCache !== null) {
```
Comment (mrow4a) [inline apps/files_sharing/tests/Command/CleanupRemoteStoragesTest.php]: the same (clear)

File: apps/testing/lib/BigFileID.php [hunk 9]
```diff
@@ -53,6 +54,9 @@ public function increaseFileIDsBeyondMax32bits(){
 												'', '4', '3', '163', '1499256550', '1499256550',
 												'0', '0', '595cd6e63f375', '27', 'NULL')";
 			$this->connection->executeQuery($query);
+			if (Cache::$metaDataCache !== null) {
```
Comment (mrow4a) [inline apps/testing/lib/BigFileID.php]: the same (clear)

File: lib/private/Files/Cache/Propagator.php [hunk 10]
```diff
@@ -96,6 +96,10 @@ public function propagateChange($internalPath, $time, $sizeDifference = 0) {
 		}
 
 		$builder->execute();
+
```
Comment (mrow4a) [inline lib/private/Files/Cache/Propagator.php]: the same (clear)

Comment (mrow4a) [reply lib/private/Files/Cache/Propagator.php]: also in other places

**Comment by butonic** [comment]
Closing in favor of OCIS.

**Pull request closed without merging**